### DISCUSSION TITLE
v2.5 is released  (Major Release)

### DISCUSSION THREAD

**Discussion by siduck** [NvChad/NvChad#2714] [Announcements]
![v2.0 poster](https://nvchad.com/news/v2.5.webp)

## Added

- Starter template for NvChad which lets users use NvChad as a plugin, thus not needing custom config stuff.

- 7 New themes for base46: 
    flexoki,
    jabuti,
    poimandres,
    mito-laser
    nano-light, 
    flexoki-light, 
    chadracula-evondev,
    Material-lighter, Material-darker, Rosepine-dawn, Solarized-oska

- Minimal terminal module which replaces nvterm which lets you created any amount of toggleable terminals , change bg color / highlights of each window etc.

- term table ( config ) in ui table of chadrc.
- Integrations list of base46 so you can exclude/include integrations supported by base46.
- Default lspconfig now exports on_init function which you can include in your custom lspconfig
- All nvchad modules will start with `nvchad` , example : require('nvchad.options')

## Changed 

- List of all defaults will be in lua/nvconfig instead of core/default_config
- Overriding modules for tabufline/statusline, its a lot easier now.
- Extended_integrations syntax for base46
- Mappings no longer use nvchad's custom syntax. It uses nvim's `vim.keymap.set`

## Removed
- Custom config structure handling.
- Nvterm plugin
- NvChad updater
- chadrc's cmp options: border_color , selected_item_bg
- tabufline, show_numbers

Read https://nvchad.com/news/v2.5_release

**Comment by MartinDelille** [comment]
The following link doesn't work: https://nvchad.com/news/v2.5

**Reply by jhonvidal** [reply]
Try this [https://nvchad.com/news/v2.5_release](https://nvchad.com/news/v2.5_release)

**Comment by le4ker** [comment]
Is it a minor release since it requires to change the code under the `custom` directory?

**Reply by siduck** [reply]
@le4ker by minor release i mean, not many features. check the migrate script. It automates 90% of the stuff. Only thing u need to do manually is updating mappings syntax to vim.keymap.set

**Reply by bcyran** [reply]
> It automates 90% of the stuff. Only thing u need to do manually is updating mappings syntax to vim.keymap.set

Not even close to true for complex config.

I don't want to come off as rude because I very much appreciate your effort put into this project and I really enjoyed using it so far. Nevertheless I think those changes should be released as a major version with huge banners warning about breaking changes and docs describing what and how changed.

Hours of config rewriting is not something I want to be surprised with when firing up my editor on Sunday morning.

**Reply by siduck** [reply]
https://github.com/NvChad/NvChad/assets/59060246/d53f5d3d-3e0a-4f37-a797-8ad3c9883c1a

@bcyran  tested with my own custom config, it works. The mapping part needs manual configuration.

**Reply by bcyran** [reply]
Yes, the script does work, it just doesn't cover all the possible cases and there are more things to fix than just mappings. In my case anyway but I would be very surprised if I'm the only one.

**Reply by siduck** [reply]
@bcyran improve the script then!

**Reply by bcyran** [reply]
I just don't think it's possible to perform such a migration with a script. Lua is a full blown programming language and neovim configs are programs, not just static files with data.

**Reply by siduck** [reply]
yes, but the script does work on most cases. I dont see why it wouldnt work, only the module paths are to be renamed.  Except for the mappings one.  For still using old mapping syntax, one could just add custom logic for it, example, looping over the whole mapping table, example:

 
![image](https://github.com/NvChad/NvChad/assets/59060246/0c02d52b-b6fe-4302-a6d0-bc927ab10a04)

**Reply by jgbmattos** [reply]
I had a few mappings with functions, so, being lazy and to conform with neovim's format, I extended the previous idea to spit a properly formatted mapping's file (to stdout):
<details>
<summary>Source code</summary>

```
local m = dofile("mappings.lua")

local filename = "mappings.lua" 
local file = io.open(filename, "r")
if not file then
    print("Could not open file: " .. filename)
    return
end
-- Add mappings.lua contents to a list to be queried later
local lines = {}

for line in file:lines() do
  table.insert(lines, line)
end
file:close()

local spc = "    "

for name, maps in pairs(m) do
  print("-- " .. name .. " mappings")
  for mode, data in pairs(maps) do
    if (mode ~= "plugin") then
      for key, val in pairs(data) do
        if type(val[1]) == "function" then
          local info = debug.getinfo(val[1], "S")
          local currLine = info.linedefined+1
          io.write("map(\""..mode.."\",\n"..spc.."\""..key.."\",\n"..spc.."function ()\n")
          repeat 
            io.write(spc..spc..lines[currLine]:match("^%s*(.-)%s*$").."\n")
            currLine = currLine + 1
          until string.find(lines[currLine], "end", 1, true) 
          io.write(spc.."end, \n"..spc.."{ desc = \""..val[2].."\"}\n)\n")
        else
          print("map(\""..mode.."\", \""..key.."\", \""..val[1].."\", { desc = \""..val[2].."\"})")
        end
      end
    end
  end
  print("")
end
```
</details>

<details>
<summary>Input/Output example</summary>

e.g.: 
```
from:
M.dap = {
  plugin = true,
  n = {
    ["<leader>db"] = {
      function ()
        require('dap').toggle_breakpoint()
      end,
      "Add breakpoint at line"
    },
  }
}
to:
-- dap mappings
map("n",
    "<leader>db",
    function ()
        require('dap').toggle_breakpoint()
    end, 
    { desc = "Add breakpoint at line"}
)
```

</details>

 PS1: This is my first lua script ever, if you know a better/cleaner way of doing it, please share :)
 PS2: I'm not sure if there's any downside in using functions in mappings. NVChad is my first contact with vim/neovim. It was working and its working now, so... please share if there' s alternatives/better way of doing it.

**Comment by xarthurx** [comment]
I somehow agrees with @bcyran 
For instance, after using the script, I got things like `You need to set 'vim.g.maplocalleader' **BEFORE** loading lazy`. And I have no clue where should I set it...

**Reply by xarthurx** [reply]
This really isn't a `minor` change, and requires the user to do quite some work.

Other things I haven't figure out are:
1. how should I set up a dotFile sync with git now, without mixing the files with default `nvChad` so that I can still update.
2. Where is the `nvchadUpdate` command? Is `Lazy sync` is good enough? Then how should I update the nvchad?

**Reply by xarthurx** [reply]
I'm going to post all the issues I encounter here so that others can find for reference:

I also modified `LuaSnip` to include my own snip folder.
The migration script changed the past settings to:
```
	{
    "L3MON4D3/LuaSnip",
    opts = {
      history = true,
      updateevents = "TextChanged,TextChangedI",
      enable_autosnippets = true,
    },
    config = function(_, opts)
      -- load default luasnip config
      require("nvchad.configs.others").luasnip(opts)

      local luasnip = require("luasnip")
      luasnip.filetype_extend("typescriptreact", { "html" })
      require("luasnip/loaders/from_vscode").lazy_load()
    end,
  },
```
But there aren't any `nvchad.config.others` modules.

**Reply by siduck** [reply]
@xarthurx   

> have no clue where should I set it...
Because such options need to be loaded before plugins as nvchad repo is being used as a plugin now.

>  dotFile sync with git now, w

just use the starter repo, delete its .git and add to your repo? or fork starter repo?  Lazy sync is enough to update nvchad

**Reply by xarthurx** [reply]
Wait, so you now put `nvchad` as a plugin? Otherwise how `Lazy sync` are going to sync core `nvchad` files?
What about in the future, the `starter-kit` also changes...

Personally I would prefer the previous structures, but I guess you have your own reason to do this.

**Reply by jhonvidal** [reply]
> I'm going to post all the issues I encounter here so that others can find for reference:
> 
> I also modified `LuaSnip` to include my own snip folder.
> The migration script changed the past settings to:
> ```
> 	{
>     "L3MON4D3/LuaSnip",
>     opts = {
>       history = true,
>       updateevents = "TextChanged,TextChangedI",
>       enable_autosnippets = true,
>     },
>     config = function(_, opts)
>       -- load default luasnip config
>       require("nvchad.configs.others").luasnip(opts)
> 
>       local luasnip = require("luasnip")
>       luasnip.filetype_extend("typescriptreact", { "html" })
>       require("luasnip/loaders/from_vscode").lazy_load()
>     end,
>   },
> ```
> But there aren't any `nvchad.config.others` modules.
> 

You can take a look at my repository, I have adapted the new version 2.5 with extended custom Luasnip configurations among others without touching the core that already imports it implicitly in nvchad.

https://github.com/jhonvidal/siuvim/blob/main/lua/plugins/init.lua

**Reply by siduck** [reply]
@xarthurx  starter kit will not change now. thats how lazyvim made this user config stuff too, their starter config makes user import lazyvim's main repo as plugin.

Now users will no need to read custom config abstraction and they can maintain their config just like standard nvim configs! in v2.0 , it'd be hard for users to git-track folders like after, ftplugin, syntax etc, cuz they were outside custom dir and they had to manage custom dir in their dotfiles repo  and it'd be a mess that they'd had to clone nvchad repo and then append their custom configs. 

With v2.5 they dont have to worry about it!   Its all clean

v2.5's codebase ( main nvchad repo ) is barely 900 loc now!

![image](https://github.com/NvChad/NvChad/assets/59060246/a83cc372-af9d-444e-9bc6-f5cb6a8f4235)

**Reply by xarthurx** [reply]
@siduck OK, after comparing and reading the code, I have more clue now...
So, if I have other plugins that are parallel with nvchad, then
```
- lua
  |- nvchad
  |- nvchadconfig
- init.lua
```

Otherwise, similar to the starterKit:
```
- lua
  |- config1
  |- config2
- init.lua
```
correct?

In this way, `localleader` stuff I mentioned above should be in the `init.lua` before `nvchad` and `lazy` is loaded.

**Reply by xarthurx** [reply]
No, the above interpretation is incorrect.
Only the starterkit structure is correct....

**Reply by siduck** [reply]
@xarthurx  yes starterkit structure is correct, nvchad repo now will be used for importing its modules aka `nvchad.options` etc

starter's structure is very simple, like if you dont want `nvchad.mappings`, dont require it. And so on!

**Reply by xarthurx** [reply]
@siduck  Thank you for the quick reply. One last question:
I'm using `Which-key` and my previous key mappings will integrate new mappings into the which-key menu.
The current new mechanism doesn't seem to load my previous which-key settings:

```
require "nvchad.mappings"

local M = {}
M.truzen = {
   n = {
      ["<leader>ta"] = { "<cmd> TZAtaraxis <CR>", "truzen ataraxis" },
      ["<leader>tm"] = { "<cmd> TZMinimalist <CR>", "truzen minimal" },
      ["<leader>tf"] = { "<cmd> TZFocus <CR>", "truzen focus" },
   },
}
```
Is there any documentation on that or how could I migrate?

**Reply by siduck** [reply]
@xarthurx  you could just use vim.keymap.set to set mappings https://nvchad.com/docs/config/mappings

 or make a wrapper function that loops over M table + each of its groups and sets mappings . ( v2.0 was doing this too, was using vim.keymap.set under the hood )
 
 So ig u dont need `core.utils.load_mappings('blabla')` too, just load all mappings at start. Except for buffer specific ones, like our  lspconfig's / gitsigns on_attach has

**Reply by AlexanderRevo** [reply]
> You need to set 'vim.g.maplocalleader' **BEFORE** loading lazy

Did you ever resolve this specific issue? What was the fix?

**Reply by xarthurx** [reply]
> > You need to set 'vim.g.maplocalleader' **BEFORE** loading lazy
> 
> Did you ever resolve this specific issue? What was the fix?

Something like this:

```lua
vim.g.base46_cache = vim.fn.stdpath("data") .. "/nvchad/base46/"
vim.g.mapleader = " "
vim.g.maplocalleader = ";"  --- this line
 
-- bootstrap lazy and all plugins
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"

if not vim.uv.fs_stat(lazypath) then
	local repo = "https://github.com/folke/lazy.nvim.git"
	vim.fn.system({ "git", "clone", "--filter=blob:none", repo, "--branch=stable", lazypath })
end

vim.opt.rtp:prepend(lazypath)

local lazy_config = require("configs.lazy")

-- load plugins
require("lazy").setup({
	{
		"NvChad/NvChad",
		lazy = false,
		branch = "v2.5",
		import = "nvchad.plugins",
	},

	{ import = "plugins" },
}, lazy_config)

```

**Reply by AlexanderRevo** [reply]
Oh, apparently, I can't read. I only had `mapleader` set, and for some reason thought "but it's right there!" :facepalm: I still have the issue, though, but it only seems to occur when opening vimwiki, so maybe it is the culprit after all, not my NvChad config.

**Reply by siduck** [reply]
@AlexanderRevo  check nvchad/starter repo!!!!!!!!!!!!!!!!!!!!!!!!

**Reply by AlexanderRevo** [reply]
> @AlexanderRevo check nvchad/starter repo!!!!!!!!!!!!!!!!!!!!!!!!

I am using the starter repo. What should I be looking for exactly?

**Reply by siduck** [reply]
@AlexanderRevo whats the issue you're facing?

**Comment by siduck** [comment]
> I'm going to post all the issues I encounter here so that others can find for reference:
> 
> I also modified `LuaSnip` to include my own snip folder. The migration script changed the past settings to:
> 
> ```
> 	{
>     "L3MON4D3/LuaSnip",
>     opts = {
>       history = true,
>       updateevents = "TextChanged,TextChangedI",
>       enable_autosnippets = true,
>     },
>     config = function(_, opts)
>       -- load default luasnip config
>       require("nvchad.configs.others").luasnip(opts)
> 
>       local luasnip = require("luasnip")
>       luasnip.filetype_extend("typescriptreact", { "html" })
>       require("luasnip/loaders/from_vscode").lazy_load()
>     end,
>   },
> ```
> 
> But there aren't any `nvchad.config.others` modules.

https://github.com/NvChad/NvChad/tree/v2.5/lua/nvchad/configs , yea others has been removed. i forgot this aspect in the script

**Comment by guzman109** [comment]
How do you call this "Minor Changes" when it completely changes the the configs structure? There is no custom directory anymore and I need to completely change all custom configs I had.

**Reply by siduck** [reply]
by minor i meant less features

> completely change all custom configs I had.

didnt you use the migration script?

**Reply by xarthurx** [reply]
@siduck  I really appreciate your work and your effort into nvChad -- this is huge.
I also release my own software.
IMHO, we cannot just define major/minor based on features... It should be based on how much it will affect the previous user.

For changes from v2.0 - v2.5, this is really not minor. For my case, I need to spend several hours fixing things here and there -- (this is the time already with your quick response. Cannot imagine for others.)

As NvChad has more and more users now, and is kind of wide spread, you as the main developer needs to be more careful about things like this...

But this is just my suggestions. You have the full right to do whatever you want as this is an open sourced software.

**Reply by siduck** [reply]
@xarthurx the migration script should work 90% of the time tho except for the mappings part, you can improve the script if you want.

Also no one's forcing y'all to use v2.5 this early, you can still use v2.0 and use nvim_appname feature to maintain v2.5 separately. 

v2.5 was a rushed release because i will probably be inactive this month to april 11 and had decided this new syntax for v3.0 but i didnt wanted new users to read the old syntax and waste their time in migrating again to new syntax so I released this in a  hurry

**Reply by guzman109** [reply]
Thanks for the reply. No I have not used or seen the migration script. I'll try it out. 

 As @xarthurx pointed out,this is not minor at all. This would have been better to be released with v3.0 as you are completely changing the structure of the system. You could have had this new structure in a nightly version and see what user feedback would have been like. This would give users a bit more time to be ready for the changes in the next major release rather than doing it in a minor release. 

The removed section in your changelog alone indicates this would be a major release.

**Reply by siduck** [reply]
@guzman109 you're right, i always thought major/minor words in releases meant for features. I removed the minor word from this post. 

> This would give users a bit more time to be ready for the changes in the next major release rather 

 you can still use v2.0, and use nvim appname to use v2.5 & slowly migrate.

**Reply by guzman109** [reply]
@siduck Thanks again for your fast replies and hard work. I don't mean to pick on you in any way, this is a lesson learned for all of us. Thankfully your docs have always been amazing and I will migrate to v2.5.

**Reply by msundin** [reply]
@siduck I recommend having a look at [Semantic Versioning](https://semver.org/) which is a widely used best practice for versioning of software with clear definitions of a major/minor/patch release.

**Comment by ftqo** [comment]
I already posted this on Reddit, but I'm copying it here as it is more relevant:

> Flexoki was not correctly ported. The light version is supposed to use the dark versions of those colors, and the dark version is supposed to use the light versions of those colors. The light version still looks nice, but it doesn't fit the purpose of the theme (to be "inky").

[See the official Flexoki website for more details.](https://stephango.com/flexoki)

**Reply by siduck** [reply]
thx for correcting.try the fix branch of base46

**Reply by ftqo** [reply]
Looks good to me. As a last note, I think some of the greys can be replaced with the greys that are also on the spec page, but I'm not too sure how that would actually look in NvChad.

**Reply by siduck** [reply]
those greys are based on flexoki's bg color, they should be in steps i.e grey would be lighter by 6-7% by grey_fg and so on , so cant change them.
